packages/core/src/combinator/withArrayValues.js

```diff
@@ -1,15 +1,16 @@
 /** @license MIT License (c) copyright 2010-2017 original author or authors */
 
-import { empty } from '../source/empty'
 import Pipe from '../sink/Pipe'
+import { empty } from '../source/empty'
+import { take } from './slice'
 
 export const withArrayValues = (array, stream) =>
   zipArrayValues(keepLeft, array, stream)
 
 export const zipArrayValues = (f, array, stream) =>
   array.length === 0 || stream === empty()
     ? empty()
-    : new ZipArrayValues(f, array, stream)
+    : new ZipArrayValues(f, array, take(array.length, stream))
```

Comment:
Did you have any thoughts about the wrapping order of take vs. zipArrayValues?  I'm not sure it matters much, but figured I'd ask.

One thing that could be interesting is to fold `empty()` handling into slice()--might be something we want to do regardless of what we decide here.

If we did that, then anything wrapped in slice() (or take, or skip) would automatically gain the empty() or zero-length slice optimization.  I don't think we could fully take advantage of it in this particular case, though, since `take(array.length, new ZipArrayValues(...))` would never trigger a `=== empty()` check inside `take`.  However, it *would* trigger a zero-length slice check--i.e. we could remove the zero-length check in zipArrayValues.

Reply:
Yeah, I'm cool with `take(0)` returning `empty()`

Reply:
See #150 for a more thinking on it.